spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkVariants.java

```diff
@@ -0,0 +1,238 @@
+    testPrimitiveVariant(Variants.of("hello world"), "hello world");
+  }
+
+  @Test
+  public void testVariantArrayRoundTrip() {
+    VariantMetadata metadata = Variants.emptyMetadata();
+    ValueArray array = Variants.array();
+    array.add(Variants.of("hello"));
+    array.add(Variants.of(42));
+    array.add(Variants.ofNull());
+
+    testVariantRoundTrip(metadata, array);
+  }
+
+  @Test
+  public void testVariantObjectRoundTrip() {
+    VariantMetadata metadata = Variants.metadata("name", "age", "active");
+    ShreddedObject object = Variants.object(metadata);
+    object.put("name", Variants.of("John Doe"));
+    object.put("age", Variants.of(30));
+    object.put("active", Variants.of(true));
+
+    testVariantRoundTrip(metadata, object);
+  }
+
+  @Test
+  public void testVariantNestedStructures() {
+    VariantMetadata metadata = Variants.metadata("user", "scores", "address", "city", "state");
+
+    // Create nested object: address
+    ShreddedObject address = Variants.object(metadata);
+    address.put("city", Variants.of("Anytown"));
+    address.put("state", Variants.of("CA"));
+
+    // Create array of scores
+    ValueArray scores = Variants.array();
+    scores.add(Variants.of(95));
+    scores.add(Variants.of(87));
+    scores.add(Variants.of(92));
+
+    // Create main object
+    ShreddedObject mainObject = Variants.object(metadata);
+    mainObject.put("user", Variants.of("Jane"));
+    mainObject.put("scores", scores);
+    mainObject.put("address", address);
+
+    testVariantRoundTrip(metadata, mainObject);
+  }
+
+  @Test
+  public void testVariantDateTimeTypes() {
+    // Test date/time variants
+    // Calculate the correct day offset for 2024-01-15
+    testPrimitiveVariant(Variants.ofIsoDate("2024-01-15"), 19737); // days since epoch (corrected)
+    testPrimitiveVariant(Variants.ofIsoTimestamptz("2024-01-15T10:30:00Z"), 1705314600000000L);
+    testPrimitiveVariant(Variants.ofIsoTimestampntz("2024-01-15T10:30:00"), 1705314600000000L);
+    testPrimitiveVariant(Variants.ofIsoTime("10:30:00"), 37800000000L);
+  }
+
+  @Test
+  public void testVariantBinaryAndUUID() {
+    // Test binary and UUID variants
+    byte[] binaryData = {1, 2, 3, 4, 5};
+    testPrimitiveVariant(Variants.of(ByteBuffer.wrap(binaryData)), ByteBuffer.wrap(binaryData));
+
+    UUID uuid = UUID.randomUUID();
+    testPrimitiveVariant(Variants.ofUUID(uuid), uuid);
+  }
+
+  @Test
+  public void testUnsupportedOperations() {
+    // Test that basic operations work
+    // This tests the current state where Spark integration is not fully implemented
+
+    // These tests document the current behavior and should be updated
+    // when full Spark integration is implemented
+
+    // This should not throw - just verify the classes exist and basic operations work
+    VariantVal sparkVariant = new VariantVal(new byte[] {0}, new byte[] {1, 0});
+    assertThat(sparkVariant.toString()).isNotNull(); // Should not throw
```

Comment:
This uses Spark's `Variant` class that does work with the variant data because it converts to JSON, but I think it would be better to have tests that check the variant structure and values are correct, rather than just asserting that the structure can be converted to JSON.

Reply:
I updated to check time/timestamp nanos are not supported now.